gradle-jdks-setup/src/main/java/com/palantir/gradle/jdks/setup/GradleJdkInstallationSetup.java

```diff
@@ -121,16 +122,37 @@ private static boolean copy(ILogger logger, Path destinationJdkInstallationDirec
                 lockFile, StandardOpenOption.READ, StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
             channel.lock();
             // double-check, now that we hold the lock
-            if (Files.exists(destinationJdkInstallationDirectory)) {
+            if (Files.exists(destinationJdkInstallationDirectory)
+                    && Files.exists(destinationJdkInstallationDirectory.resolve("bin/java"))) {
                 logger.log(String.format("Distribution URL %s already exists", destinationJdkInstallationDirectory));
                 return false;
             }
             logger.log(
                     String.format("Copying JDK from %s into %s", currentJavaHome, destinationJdkInstallationDirectory));
-            FileUtils.copyDirectory(currentJavaHome, destinationJdkInstallationDirectory);
+
+            // same filesystem for the temporary diretory as the destinationDirectory to ensure an atomic move
+            Path tempCopyDir = jdksInstallationDirectory.resolve(
+                    String.format("tmp-%s", destinationJdkInstallationDirectory.getFileName()));
+            // Ensuring that the JDK installation directory won't be partially copied.
+            // 1. Copying the currentJavaHome to a temporary directory inside the jdksInstallationDirectory
+            // 2. Atomic move of the temporary directory to the destination directory
+            try {
+                Files.createDirectories(tempCopyDir);
+                FileUtils.copyDirectory(currentJavaHome, tempCopyDir);
+                Files.move(
+                        tempCopyDir,
+                        destinationJdkInstallationDirectory,
+                        StandardCopyOption.REPLACE_EXISTING,
+                        StandardCopyOption.ATOMIC_MOVE);
```

Comment:
The common filesystems (ieg APFS/NTFS/ext4) all support atomic moves... but perhaps we should handle the `AtomicMoveNotSupportedException` like before just incase this ends up on a different filesystem that doesn't?